src/QEQ/fix_qeq_ctip.cpp

```diff
@@ -0,0 +1,394 @@
+  }
+
+  if (i > maxrepeat && comm->me == 0)
+    error->all(FLERR,"Fix qeq some charges not bound within the domain");
+
+  if (force->kspace) force->kspace->qsum_qsq();
+}
+
+/* ---------------------------------------------------------------------- */
+
+void FixQEqCTIP::init_matvec()
+{
+  compute_H();
+
+  int inum, ii, i;
+  int *ilist;
+  double *q = atom->q, qi;
+
+  double r = cutoff;
+  double rsq = r*r;
+  double r6 = rsq*rsq*rsq;
+  
+  double erfcd_cut = exp(-cdamp * cdamp * rsq);
+  double t_cut = 1.0 / (1.0 + EWALD_P * cdamp * r);
+  double erfcc_cut = (t_cut * (A1 + t_cut * (A2 + t_cut * (A3 + t_cut * (A4 + t_cut * A5)))) * erfcd_cut) / r;
+
+  int elt1;
+  double reff[atom->ntypes], reffsq[atom->ntypes], reff4[atom->ntypes], reff7[atom->ntypes], s2d_self[atom->ntypes];
+
+  for (elt1 = 0; elt1 < atom->ntypes; elt1++) {
+    reff[elt1] = cbrt(rsq*r + 1/(gamma[elt1+1]*gamma[elt1+1]*gamma[elt1+1]));
+    reffsq[elt1] = reff[elt1]*reff[elt1];
+    reff4[elt1] = reffsq[elt1]*reffsq[elt1];
+    reff7[elt1] = reff4[elt1]*reffsq[elt1]*reff[elt1];
+    s2d_self[elt1] = 2.0*force->qqr2e*(1.5*erfcc_cut + 2.0*cdamp/MY_PIS*erfcd_cut + cdamp*cdamp*cdamp/MY_PIS*rsq*erfcd_cut + 0.5/reff[elt1] - 1.5/r + r6/reff7[elt1] + cdamp/MY_PIS);
+  }
+
+  inum = list->inum;
+  ilist = list->ilist;
+
+  for (ii = 0; ii < inum; ++ii) {
+    i = ilist[ii];
+    if (atom->mask[i] & groupbit) {
+      
+      qi=q[i];
+      if (qi < qmin[atom->type[i]]) {
+        Hdia_inv[i] = 1. / (eta[atom->type[i]]+2*omega[atom->type[i]]-s2d_self[atom->type[i]-1]);
+        b_s[i]      = -((chi[atom->type[i]]-2*qmin[atom->type[i]]*omega[atom->type[i]]) + chizj[i]);
+      } else if (qi < qmax[atom->type[i]]) {
+        Hdia_inv[i] = 1. / (eta[atom->type[i]]-s2d_self[atom->type[i]-1]);
+        b_s[i]      = -(chi[atom->type[i]] + chizj[i]);
+      } else {
+        Hdia_inv[i] = 1. / (eta[atom->type[i]]+2*omega[atom->type[i]]-s2d_self[atom->type[i]-1]);
+        b_s[i]      = -((chi[atom->type[i]]-2*qmax[atom->type[i]]*omega[atom->type[i]]) + chizj[i]);
+      }
+      
+      b_t[i]      = -1.0;
+      t[i] = t_hist[i][2] + 3 * (t_hist[i][0] - t_hist[i][1]);
+      s[i] = 4*(s_hist[i][0]+s_hist[i][2])-(6*s_hist[i][1]+s_hist[i][3]);
+    }
+  }
+
+  pack_flag = 2;
+  comm->forward_comm(this); //Dist_vector(s);
+  pack_flag = 3;
+  comm->forward_comm(this); //Dist_vector(t);
+}
+
+/* ---------------------------------------------------------------------- */
+
+void FixQEqCTIP::compute_H()
+{
+  int inum, jnum, *ilist, *jlist, *numneigh, **firstneigh;
+  int i, j, ii, jj;
+  double **x;
+  double dx, dy, dz, r_sqr, r, reff;
+  double cutoffsq, cutoffcu, cutoff4, cdampcu, erfcc_cut, erfcd_cut, t_cut;
+  double erfcc, erfcd, t;
+  int elt1, elt2;
+  double shield[atom->ntypes][atom->ntypes], shieldcu[atom->ntypes][atom->ntypes], reffc[atom->ntypes][atom->ntypes], reffcsq[atom->ntypes][atom->ntypes], reffc4[atom->ntypes][atom->ntypes], reffc7[atom->ntypes][atom->ntypes];
```

Comment:
Variable length arrays (VLAs) are not valid C++ (it's a g++ extension). Please avoid.

Reply:
@gplummer317 This *must* be corrected. Otherwise, we cannot merge your pull request, as it will no pass all tests.

For the pair style, you can just move the arrays as pointer variables into the header and allocate them with memory->create() and copy over the data in the setup_params() function (make certain to initialized the pointers to nullptr in the constructor and first destroy() and then create() them and also destroy() them in the destructor). This should give you a (small?) performance improvement, since you allocate and initialize them only once per run and not once for every call to compute().

For the fix a similar solution needs to be found. I've made some formal/cosmetic updates and merged with upstream.

Reply:
fixed